jme3-android-native/src/native/jme_bufferallocator/com_jme3_util_AndroidNativeBufferAllocator.c

```diff
@@ -0,0 +1,85 @@
+ *   without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file com_jme3_util_AndroidNativeBufferAllocator.c
+ * @author pavl_g.
+ * @brief Creates and releases direct byte buffers for {com.jme3.util.AndroidNativeBufferAllocator}.
+ * @date 2022-05-17.
+ * @note
+ * Find more at :
+ * - JNI Direct byte buffers : https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#NewDirectByteBuffer.
+ * - JNI Get Direct byte buffer : https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#GetDirectBufferAddress.
+ * - GNU Allocating memory : https://www.gnu.org/software/libc/manual/html_node/Basic-Allocation.html.
+ * - GNU Freeing memory : https://www.gnu.org/software/libc/manual/html_node/Freeing-after-Malloc.html.
+ * - Android logging : https://developer.android.com/ndk/reference/group/logging.
+ * - Android logging example : https://github.com/android/ndk-samples/blob/7a8ff4c5529fce6ec4c5796efbe773f5d0e569cc/hello-libs/app/src/main/cpp/hello-libs.cpp#L25-L26.
+ */
+
+#include "headers/com_jme3_util_AndroidNativeBufferAllocator.h"
+#include <stdlib.h>
+
+#ifndef NDEBUG
+#include <android/log.h>
+#define LOGI(...) __android_log_print(ANDROID_LOG_INFO, \
+                                     "AndroidNativeBufferAllocator", ##__VA_ARGS__);
+#else
+#define LOGI(...)
+#endif
+
+JNIEXPORT void JNICALL Java_com_jme3_util_AndroidNativeBufferAllocator_releaseDirectByteBuffer
+(JNIEnv * env, jobject object, jobject bufferObject)
+{
+    void* buffer = (*env)->GetDirectBufferAddress(env, bufferObject);
+    // deallocates the buffer pointer
+    free(buffer);
+    // log the destruction by mem address
+    LOGI("Buffer released (mem_address, size) -> (%p, %lu)", buffer, sizeof(buffer));
+    // avoid accessing this memory space by resetting the memory address
+    buffer = NULL;
+    LOGI("Buffer mem_address formatted (mem_address, size) -> (%p, %lu)", buffer, sizeof(buffer));
+}
+
+JNIEXPORT jobject JNICALL Java_com_jme3_util_AndroidNativeBufferAllocator_createDirectByteBuffer
+(JNIEnv * env, jobject object, jlong size)
+{
+    void* buffer = malloc(size);
```

Comment:
The jvm specification requires buffers to be zeroed.
We should either have a memset that sets the memory to 0s or use calloc instead of malloc here.

Reply:
I am not sure if i am fully understanding you, so we will do these steps after the basic `malloc(size_t size)`: 

- if (`buffer != 0`) then ` memset (buffer, 0, size);`.
- OR just replace the whole thing with a single line : `void* buffer = calloc (1, size)`.

Which one do you think is better ?

Reply:
I suggest using calloc from stdlib, much cleaner and reliable.

Reply:
Yes, calloc.

In my hardcode C programming days, malloc was always looked on with suspicion... too often where malloc goes so does memory bugs.

Reply:
Alright, i will go with `calloc(1, size)`, thanks.

Reply:
Oh i forgot this part, i think we should implement our own `calloc()`, so we can check if the malloc is zero and this means the memory is exhausted and the program should throw a fatal error (exiting with ENOMEM error code) describing that the device is out of memory :

```c
#include <errno.h>

void* xcalloc (size_t size) {
  void* buffer = malloc (size);
  // sanity check the buffer
  if (buffer == 0) {
     LOGE("Error device is out of memory, cannot allocate (%p, %ul), buffer, size);
     LOGE("Exiting with error code %i ", errno);
     exit(errno);
   } else {
     memset (buffer, 0, size);
     LOGI("Buffer created successfully (%p, %ul)", buffer, size);
   }
  return value;
}
```
Are we okay with this change ? 

Generally, combining those : 
- https://www.gnu.org/software/libc/manual/html_node/Allocating-Cleared-Space.html
- https://www.gnu.org/software/libc/manual/html_node/Malloc-Examples.html

Reply:
Pretty classic problem, though... how are you going to log anything if you can't allocate?

...why switch back to malloc, though?

Reply:
> Pretty classic problem, though... how are you going to log anything if you can't allocate?
> 
> ...why switch back to malloc, though?

1) This was an example from GNU C docs...but it was using fatal(..) which is printf or perror, so could use perror instead but i am not sure if it will show the error for jni.

2) I haven't switched back to malloc, instead i implemented calloc locally as stated by GNU docs : 
```c
You could define calloc as follows:

void *
calloc (size_t count, size_t eltsize)
{
  void *value = reallocarray (0, count, eltsize);
  if (value != 0)
    memset (value, 0, count * eltsize);
  return value;
}
```

Reply:
The code allocates memory only here, so I think there is no reason to create the xcalloc function since checks can be done here in the only place where they are needed.
Also i think calloc should be used anyway, since the code you posted is only one possible implementation, platforms could implement it differently in their stdlib

Reply:
Yes, optimal to use calloc instead of rolling your own probably-degraded version.

Reply:
I just want to handle memory errors in some logs, suggest me a better solution if you know, something like, may be `perror()` is sufficient : 

[Code] : 
```c
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>

void* xcalloc (size_t size) {
  void* buffer = malloc (size);
  // sanity check the buffer
  if (buffer == 0) {
     perror("Exiting with error code");
     exit(errno);
   } else {
     memset (buffer, 0, size);
     printf("Buffer created successfully (%p, %lu)", buffer, size);
   }
  return buffer;
}

int main() {
    void* buffer = xcalloc(50000000000000000*50000000000000000);
    return 0;
}

```
[Debug logs] : 
```c
Exiting with error code: Cannot allocate memory
```

Reply:
> The code allocates memory only here

I haven't tested what will happen if the device is out of memory (eg. allocating multiple large buffers at the same time)....the crash should be obvious that is the application is halted because we are out of memory...but at least from my opinion we should handle memory errors properly for future debugging purposes, of note, emulators are small memory virtual devices so expect to have these kinds of problems especially on emulators.

Reply:
I think the logging is fine, this will be probably  triggered when the app tries to allocate a very large memory area, so there will probably be enought memory available for the logging.

Reply:
Alright, i think this may be better if you insist to use calloc, what do you think ? : 
```c
#include <errno.h>
#include <stdlib.h>
#include <stdbool.h>

bool isDeviceOutOfMem(void* buffer) {
   return buffer != NULL && errno == ENOMEM;
}

void* buffer = calloc(1, size);
if (isDeviceOutOfMem(buffer)) { 
   perror("Device is out of memory");
   exit(errno);
} else { 
  LOGI("Buffer created successfully %p %ul ", buffer, size);
}
```

Reply:
Tested this and the perror shows nothing on android, `ANDROID_LOG_FATAL` should be used instead, i will implement this approach if there is no objection.

Reply:
This what the test shows, on regular run : 
```java
....
I/AndroidNativeBufferAllocator: Buffer released (mem_address, size) -> (0x78c08ab320, 8)
    Buffer mem_address formatted (mem_address, size) -> (0x0, 8)
    Buffer created successfully (mem_address, size) -> (0x78d0948130 65536)
    Buffer released (mem_address, size) -> (0x78c08c3950, 8)
    Buffer mem_address formatted (mem_address, size) -> (0x0, 8)
    Buffer created successfully (mem_address, size) -> (0x78d0938120 65536)
I/AndroidNativeBufferAllocator: Buffer released (mem_address, size) -> (0x78c08b3530, 8)
....
```

When setting buffer = NULL and errno = ENOMEM explicitly to mimic no memory : 
```java
I/JmeSurfaceView: Starting a new Game State
I/JmeSystem: Running on jMonkeyEngine 3.4.0-SNAPSHOT
     * Branch: android-native-bytebuffer
     * Git Hash: f46d469
     * Build Date: 2022-05-21
I/OGLESContext: Android Build Version: {0}
I/JmeSurfaceView: Game returns from the idle mode
I/JmeSurfaceView: Game goes idle
I/JmeSurfaceView: Context stops, but game is still running
I/JmeSurfaceView: Starting a new Game State
I/OGLESContext: Android Build Version: {0}
I/JmeSurfaceView: Game returns from the idle mode
I/AdrenoGLES: QUALCOMM build                   : f2ab992, I401605978b
    Build Date                       : 09/28/19
    OpenGL ES Shader Compiler Version: EV031.27.05.01
    Local Branch                     : 
    Remote Branch                    : 
    Remote Branch                    : 
    Reconstruct Branch               : 
    Build Config                     : S L 8.0.11 AArch64
I/AdrenoGLES: PFP: 0x005ff11a, ME: 0x005ff066
I/Gralloc4: mapper 4.x is not supported
W/Gralloc3: mapper 3.x is not supported
I/JmeSurfaceView: JmeSurfaceView's joined the UI thread
I/JmeSurfaceView: JmeSurfaceView's joined the UI thread
A/AndroidNativeBufferAllocator: Device is out of memory
```